File: scripts/algorithms/M/Maximum Product of Word Lengths/Maximum Product of Word Lengths.java
// Runtime: 14 ms (Top 91.40%) | Memory: 44.6 MB (Top 93.55%)
class Solution {
    public int maxProduct(String[] words) {
        int n = words.length;
        int[] masks = new int[n];

        for (int i=0; i<n; i++)
            for (char c: words[i].toCharArray())
                masks[i] |= (1 << (c - 'a'));

        int largest = 0;
        for (int i=0; i<n-1; i++)
            for (int j=i+1; j<n; j++)
                if ((masks[i] & masks[j]) == 0)
                    largest = Math.max(largest, words[i].length() * words[j].length());

        return largest;
    }
}

File: scripts/algorithms/M/Maximum Product of Word Lengths/Maximum Product of Word Lengths.rs
// Runtime: 7 ms (Top 100.0%) | Memory: 4.00 MB (Top 66.67%)

impl Solution {
    pub fn max_product(words: Vec<String>) -> i32 {
        let v = words
            .iter()
            .map(|word| {
                word.as_bytes()
                    .iter()
                    .fold(0_u32, |acc, x| acc | 1 << (x - b'a'))
            })
            .collect::<Vec<_>>();
        let mut answer = 0;
        for i in 0..words.len() - 1 {
            for j in i + 1..words.len() {
                if v[i] & v[j] == 0 {
                    answer = answer.max(words[i].len() * words[j].len())
                }
            }
        }
        answer as i32
    }
}
